frame/nomination-pools/src/migration.rs

```diff
@@ -384,3 +384,67 @@ pub mod v2 {
 		}
 	}
 }
+
+pub mod v3 {
+	use super::*;
+
+	/// This migration removes stale bonded-pool metadata, if any.
+	pub struct MigrateToV3<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV3<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let current = Pallet::<T>::current_storage_version();
+			let onchain = Pallet::<T>::on_chain_storage_version();
+
+			log!(
+				info,
+				"Running migration with current storage version {:?} / onchain {:?}",
+				current,
+				onchain
+			);
+
+			if current > onchain {
+				let mut metadata_iterated = 0u64;
+				let mut metadata_removed = 0u64;
+				Metadata::<T>::iter_keys()
+					.filter(|id| {
+						metadata_iterated += 1;
+						!BondedPools::<T>::contains_key(&id)
+					})
+					.collect::<Vec<_>>()
+					.iter()
+					.for_each(|id| {
+						metadata_removed += 1;
+						Metadata::<T>::remove(&id);
+					});
```

Comment:
```suggestion
					.into_iter()
					.for_each(|id| {
						metadata_removed += 1;
						Metadata::<T>::remove(&id);
					});
```

For the sake of posterity, I am not sure if this collect was needed, or doing `remove` inside the iteration was bad or not 🤷  But yeah better safe than sorry.